admin/services/data-access/headteacher-declaration.data.service.js

```diff
@@ -49,4 +49,98 @@ headteacherDeclarationDataService.sqlFindHdfForCheck = async (dfeNumber, checkWi
+    SELECT COUNT(p.id) as pupilsCount
+    FROM [mtc_admin].[pupil] p
+    JOIN [mtc_admin].[school] s ON p.school_id = s.id
+    LEFT JOIN [mtc_admin].[pupilStatus] ps ON (p.pupilStatus_id = ps.id)
+    WHERE s.dfeNumber = @dfeNumber
+    AND ps.code NOT IN ('NOT_TAKING', 'COMPLETED')
+  `
+
+  const params = [
+    { name: 'dfeNumber', type: TYPES.Int, value: dfeNumber }
+  ]
+
+  const result = await sqlService.query(sql, params)
+  return R.path(['pupilsCount'], R.head(result))
+}
+
+/**
+ * Find count of pupils blocking hdf submission before check end date
+ * @param dfeNumber
+ * @return {Number}
+ */
+headteacherDeclarationDataService.sqlFindPupilsBlockingHdfAfterCheckEndDate = async (dfeNumber) => {
+  const sql = `
+    SELECT COUNT(p.id) as pupilsCount
+    FROM [mtc_admin].[pupil] p
+    JOIN [mtc_admin].[school] s ON p.school_id = s.id
+    LEFT JOIN [mtc_admin].[pupilStatus] ps ON (p.pupilStatus_id = ps.id)
+    LEFT JOIN (
+        SELECT *,
+            ROW_NUMBER() OVER (PARTITION BY pupil_id ORDER BY id DESC) as rank
+        FROM [mtc_admin].[check]
+        WHERE isLiveCheck = 1
+           ) lastCheck ON (lastCheck.pupil_id = p.id)
+    LEFT JOIN [mtc_admin].[checkStatus] cs ON (lastCheck.checkStatus_id = cs.id)
+    WHERE s.dfeNumber = @dfeNumber
+    AND (lastCheck.rank = 1 or lastCheck.rank IS NULL)
+    AND (ps.code != 'STARTED' OR cs.code != 'NTR')
+    AND ps.code NOT IN ('NOT_TAKING', 'COMPLETED') 
+  `
+
+  const params = [
+    { name: 'dfeNumber', type: TYPES.Int, value: dfeNumber }
+  ]
+
+  const result = await sqlService.query(sql, params)
+  return R.path(['pupilsCount'], R.head(result))
+}
+
+/**
+ * Fetch all pupils for a school by dfeNumber with their status codes and attendance reasons
+ * @param dfeNumber
+ * @returns {Promise<*>}
+ */
+headteacherDeclarationDataService.sqlFindPupilsWithStatusAndAttendanceReasons = async function (dfeNumber) {
+  const paramDfeNumber = { name: 'dfeNumber', type: TYPES.Int, value: dfeNumber }
+
+  const sql = `
+    SELECT
+       p.*,
+       ps.code AS pupilStatusCode,
+       cs.code AS checkStatusCode,
+       pg.group_id,
+       ac.reason,
+       ac.code as reasonCode
+    FROM [mtc_admin].pupil p
+    INNER JOIN [mtc_admin].school s
+        ON s.id = p.school_id
+    LEFT JOIN [mtc_admin].pupilStatus ps
+        ON p.pupilStatus_id = ps.id
+    LEFT JOIN [mtc_admin].[pupilAttendance] pa
+        ON p.id = pa.pupil_id AND (pa.isDeleted IS NULL OR pa.isDeleted = 0)
+    LEFT JOIN [mtc_admin].[attendanceCode] ac
+        ON pa.attendanceCode_id = ac.id
+    LEFT JOIN [mtc_admin].[pupilGroup] pg
+        ON pg.pupil_id = p.id
+    LEFT JOIN (
+        SELECT *,
+            ROW_NUMBER() OVER (PARTITION BY pupil_id ORDER BY id DESC) as rank
+        FROM [mtc_admin].[check]
+        WHERE isLiveCheck = 1
+       ) lastCheck ON (lastCheck.pupil_id = p.id)
+    LEFT JOIN [mtc_admin].[checkStatus] cs ON (lastCheck.checkStatus_id = cs.id)
+    WHERE s.dfeNumber = @dfeNumber
```

Comment:
This is a large query.  As the query is not actually outputting anything from the `school` table perhaps you could remove the school table completely and use `p.school_id` instead of `dfeNumber` to limit pupils to a certain school?

Reply:
@jon-shipley you are absolutely right, done.